File: content/snippets/go/design_patterns/adapter.go
package main

import "fmt"

type Printer interface {
	Print(message string)
}

type JSONLogger struct{}

func (JSONLogger) WriteJSON(payload string) {
	fmt.Println(payload)
}

type JSONLoggerAdapter struct {
	logger JSONLogger
}

func (a JSONLoggerAdapter) Print(message string) {
	a.logger.WriteJSON(fmt.Sprintf("{\"message\":\"%s\"}", message))
}

func main() {
	printer := JSONLoggerAdapter{logger: JSONLogger{}}
	printer.Print("build finished")
}

File: content/snippets/cpp/design_patterns/adapter.cpp
#include <iostream>
#include <string>

struct Printer {
    virtual ~Printer() = default;
    virtual void print(const std::string& message) const = 0;
};

class JsonLogger {
public:
    void writeJson(const std::string& payload) const {
        std::cout << payload << '\n';
    }
};

class JsonLoggerAdapter : public Printer {
public:
    explicit JsonLoggerAdapter(const JsonLogger& logger) : logger_(logger) {}

    void print(const std::string& message) const override {
        logger_.writeJson("{\"message\":\"" + message + "\"}");
    }

private:
    const JsonLogger& logger_;
};

int main() {
    JsonLogger logger;
    JsonLoggerAdapter printer(logger);
    printer.print("build finished");
}
